src/main/java/io/pinecone/PineconeConnection.java

```diff
@@ -113,15 +112,16 @@ private VectorServiceGrpc.VectorServiceBlockingStub applyDefaultBlockingStubConf
                 .withMaxOutboundMessageSize(DEFAULT_MAX_MESSAGE_SIZE);
     }
 
-    private static String getEndpoint(PineconeClientConfig clientConfig, PineconeConnectionConfig connectionConfig) {
-        String endpoint = String.format("%s-%s.svc.%s.pinecone.io",
-                connectionConfig.getIndexName(),
-                clientConfig.getProjectName(),
-                clientConfig.getEnvironment());
+    static String getEndpoint(PineconeClientConfig clientConfig, PineconeConnectionConfig connectionConfig) {
+        String endpoint = (connectionConfig.getConnectionUrl() != null) ?
```

Comment:
Use connection url when entered by the user, otherwise follow the existing flow of building it using environment, projectName, and indexName.